File: src/python/mpi/scatter_and_gather.py
"""
Scatter and Gather Operations

This example demonstrates MPI Scatter and Gather, which are fundamental
collective operations for distributing and collecting data.

Key concepts:
- comm.Scatter: Root distributes equal chunks of data to all processes
- comm.Gather: Root collects equal chunks of data from all processes
- Uses numpy arrays for efficient data transfer (uppercase methods)
- Each process receives/sends an equal portion

Scatter divides an array into equal parts and sends one part to each process.
Gather collects data from all processes into a single array at root.

Run: mpirun -np 4 python scatter_and_gather.py
"""

from mpi4py import MPI
import numpy as np


def main():
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    root = 0
    elements_per_process = 3

    # Root process creates the data to scatter
    if rank == root:
        # Create array with size * elements_per_process elements
        total_elements = size * elements_per_process
        data = np.arange(total_elements, dtype='i')
        print(f"Root process: Original data = {data}")
        print(f"Root process: Scattering {elements_per_process} elements to each of {size} processes")
        print()
    else:
        data = None

    # Prepare receive buffer for scattered data
    recv_buffer = np.empty(elements_per_process, dtype='i')

    # Scatter: distribute chunks to all processes
    comm.Scatter(data, recv_buffer, root=root)
    print(f"Process {rank}: Received {recv_buffer}")

    # Each process performs some computation on its data
    recv_buffer = recv_buffer * 2
    print(f"Process {rank}: Processed (x2) = {recv_buffer}")

    # Prepare gather buffer on root
    if rank == root:
        gathered_data = np.empty(size * elements_per_process, dtype='i')
    else:
        gathered_data = None

    # Gather: collect results back to root
    comm.Gather(recv_buffer, gathered_data, root=root)

    # Root displays the gathered results
    if rank == root:
        print()
        print(f"Root process: Gathered data = {gathered_data}")


if __name__ == "__main__":
    main()

File: src/cpp/mpi/scatter_and_gather.cpp
#include <mpi.h>
#include <iostream>
#include <vector>

int main(int argc, char** argv) {
    // Initialize the MPI environment
    MPI_Init(&argc, &argv);

    // Get the rank (ID) of the current process
    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    // Get the total number of processes
    int size;
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    // Define the root process
    int root = 0;

    // Create an array to be scattered by the root process
    std::vector<int> data;
    if (rank == root) {
        data.resize(size * 2);
        for (int i = 0; i < size * 2; ++i) {
            data[i] = i;
        }
        std::cout << "Process " << rank << " has data: ";
        for (const auto& val : data) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }

    // Prepare a buffer for receiving the scattered data
    std::vector<int> recvbuf(2);

    // Scatter the data from the root process to all processes
    MPI_Scatter(data.data(), 2, MPI_INT, recvbuf.data(), 2, MPI_INT, root, MPI_COMM_WORLD);
    std::cout << "Process " << rank << " received data: ";
    for (const auto& val : recvbuf) {
        std::cout << val << " ";
    }
    std::cout << std::endl;

    // Perform some operation on the received data
    // (For example, multiply each element by 2)
    for (auto& val : recvbuf) {
        val *= 2;
    }
    std::cout << "Process " << rank << " processed data: ";
    for (const auto& val : recvbuf) {
        std::cout << val << " ";
    }
    std::cout << std::endl;

    // Gather the processed data back to the root process
    std::vector<int> gathered_data;
    if (rank == root) {
        gathered_data.resize(size * 2);
    }

    MPI_Gather(recvbuf.data(), 2, MPI_INT, gathered_data.data(), 2, MPI_INT, root, MPI_COMM_WORLD);

    // The root process prints the gathered data
    if (rank == root) {
        std::cout << "Process " << rank << " gathered data: ";
        for (const auto& val : gathered_data) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }

    // Finalize the MPI environment
    MPI_Finalize();
    return 0;
}
